pilot/adapter/config/file/config.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"io/ioutil"
+	"istio.io/istio/pilot/model"
+	"fmt"
+)
+
+const (
+	defaultNamespace = "default"
+	defaultDomain    = "cluster.local"
+)
+
+// Information for a single element of configuration stored in a file.
+type ConfigRef interface {
+	Meta() *model.ConfigMeta
+	FilePath() string
+}
+
+type configRef struct {
+	meta     *model.ConfigMeta
+	filePath string
+}
+
+func (r *configRef) Meta() *model.ConfigMeta {
+	return r.meta
+}
+
+func (r *configRef) FilePath() string {
+	return r.filePath
+}
+
+// Creates a new ConfigRef with the given properties.
+func NewConfigRef(schemaType string, name string, namespace string, domain string, filePath string) ConfigRef {
```

Comment:
I think `NewConfigRef(meta *model.ConfigMeta, path string)` is a nicer API - all string parameters is easy to get wrong.

<hr>

As a golang aside, for consecutive parameters of the same type, the type can be elided and included only once at the end. In other words,
```go
NewConfigRef(schemaType string, name string, namespace string, domain string, filePath string)
```
and
```go
NewConfigRef(schemaType, name, namespace, domain, filePath string)
```
are equivalent.

Reply:
Yeah, I don't think we need a `NewXXX` for config at all, given your idea for `WithDefaults` below.